### PR TITLE
PAYARA-3434 optimise FileArchive

### PR DESCRIPTION
**Pull request by Cousjava**

### CODE REVIEW

File: nucleus/deployment/common/src/main/java/com/sun/enterprise/deploy/shared/FileArchive.java [hunk 1]
```diff
@@ -138,6 +139,12 @@
      * be sure that the staleFileManager field has been set.
      */
     private boolean isOpenedOrCreated = false;
+    
+    /**
+     * Cache for {@link #entries()}. This is done as a static variable instead of a service
+     * because FileArchive is commonly created by constructor rather than through injection.
+     */
+    private static Map<Integer, Enumeration> entriesCache = new HashMap<Integer, Enumeration>();
```
Comment (svendiedrichsen) [inline nucleus/deployment/common/src/main/java/com/sun/enterprise/deploy/shared/FileArchive.java]: You might want to consider a `java.util.WeakHashMap` to allow garbage collection to reclaim memory when needed.

File: nucleus/deployment/common/src/main/java/com/sun/enterprise/deploy/shared/FileArchive.java [hunk 2]
```diff
@@ -575,14 +589,14 @@ private boolean deleteDir(File directory) throws IOException {
          */
         if (!Files.isSymbolicLink(directory.toPath())) {
             File[] entries = directory.listFiles();
-            for (int i = 0;entries != null && i < entries.length;i++) {
-                if (entries[i].isDirectory()) {
-                    allDeletesSucceeded &= deleteDir(entries[i]);
+            for (File entrie : entries) {
```
Comment (svendiedrichsen) [inline nucleus/deployment/common/src/main/java/com/sun/enterprise/deploy/shared/FileArchive.java]: Typo: entrie -> entry

Sorry for being pedantic.

File: nucleus/deployment/common/src/main/java/com/sun/enterprise/deploy/shared/FileArchive.java [hunk 3]
```diff
@@ -799,15 +823,15 @@ public static void markDeletedArchive(final File archiveFile) {
...
 
-                final URI archiveURI = archiveFile.toURI();
+                final Path archivePath = archiveFile.toPath();
                 PrintStream ps = null;
                 try {
                     ps = new PrintStream(markerFile(archiveFile));
                 } catch (FileNotFoundException ex) {
                     throw new RuntimeException(ex);
                 }
                 for ( ; staleFileIt.hasNext(); ) {
```
Comment (svendiedrichsen) [inline nucleus/deployment/common/src/main/java/com/sun/enterprise/deploy/shared/FileArchive.java]: How about moving this for loop into the above try block and use a try-with-resources to close the PrintStream automatically?

File: nucleus/deployment/common/src/main/java/org/glassfish/deployment/common/DeploymentContextImpl.java [hunk 4]
```diff
@@ -678,6 +678,7 @@ public void postDeployClean(boolean isFinalClean) {
                 for (String className : classNamesToClean) {
                     transientAppMetaData.remove(className);
                 }
+                com.sun.enterprise.deploy.shared.FileArchive.clearCache();
```
Comment (svendiedrichsen) [inline nucleus/deployment/common/src/main/java/org/glassfish/deployment/common/DeploymentContextImpl.java]: Shouldn't clearCache be called when isFinalClean is true?

File: nucleus/deployment/common/src/main/java/com/sun/enterprise/deploy/shared/FileArchive.java (hunk 1 above)
Comment (Cousjava) [reply nucleus/deployment/common/src/main/java/com/sun/enterprise/deploy/shared/FileArchive.java]: I considered WeakHashMap but that causes the entry to be collected when the key is no longer used, and the keys are a hashcode of the file. The same file is may have more than one instance of File referring to it, but the hashcode will be the same for them, which is the purpose of this cache.

Comment (svendiedrichsen) [reply nucleus/deployment/common/src/main/java/com/sun/enterprise/deploy/shared/FileArchive.java]: Ok, got it.

File: nucleus/deployment/common/src/main/java/com/sun/enterprise/deploy/shared/FileArchive.java [hunk 5]
```diff
@@ -296,12 +303,19 @@ public boolean isDirectory(String name) {
...
-        getListOfFiles(file, namesList, null);
-        return Collections.enumeration(namesList);
+        
+        if (entriesCache.containsKey(file.hashCode())) {
+            return entriesCache.get(file.hashCode());
+        } else {
+            List<String> namesList = new ArrayList<String>();
+            getListOfFiles(file, namesList, null);
+            Enumeration entries = Collections.enumeration(namesList);
+            entriesCache.put(file.hashCode(), entries);
```
Comment (svendiedrichsen) [inline nucleus/deployment/common/src/main/java/com/sun/enterprise/deploy/shared/FileArchive.java]: Looks like you are doing the job of a HashMap explicitly. Wouldn't it be sufficient do put the file in as key?

**Comment by Cousjava** [comment]
Jenkins test please

**Comment by Cousjava** [comment]
Jenkins test please

**Comment by Cousjava** [comment]
Jenkins test please

**Comment by MarkWareham** [comment]
jenkins test please

**Comment by MarkWareham** [comment]
failure due to something else

**Comment by smillidge** [comment]
jenkins test please

**Comment by Cousjava** [comment]
Test failure is due to CDI Async test, no idea why as this PR does not touch CDI

**Comment by arjantijms** [comment]
jenkins test please

**Comment by MarkWareham** [comment]
Test problem tracked with QA-154 Jira

**Comment by Cousjava** [comment]
Jenkins test please

**Pull request merged**